src/mtype.d

```diff
@@ -10247,12 +10259,70 @@ extern (C++) final class Parameter : RootObject
     final bool isCovariant(const Parameter p) const pure nothrow @nogc @safe
     {
         enum stc = STCref | STCin | STCout | STClazy;
-        return !((this.storageClass & stc) != (p.storageClass & stc) ||
+        if ((this.storageClass & stc) != (p.storageClass & stc))
+            return false;
+        return isCovariantScope(this.storageClass, p.storageClass);
+    }
+
+    static bool isCovariantScope(StorageClass from, StorageClass to) pure nothrow @nogc @safe
+    {
+        if (from == to)
+            return true;
+
+        enum
+        {
+            Mnone,
+            Mscope,
+            Mreturn,
+            MreturnScope
+        }
 
-                // We can add scope, but not subtract it
-                (!(this.storageClass & STCscope) && (p.storageClass & STCscope)) ||
+        /* Shrinking the representation is necessary because StorageClass is so wide
+         */
+        static uint buildM(StorageClass stc) pure nothrow @nogc @safe
+        {
+            uint result;
+            final switch (stc & (STCscope | STCreturn))
+            {
+                case 0:                    result = Mnone;        break;
+                case STCscope:             result = Mscope;       break;
+                case STCreturn:            result = Mreturn;      break;  // i.e. 'return ref'
+                case STCscope | STCreturn: result = MreturnScope; break;
+            }
+            return result;
+        }
 
-                // We can subtract return, but not add it
-                ((this.storageClass & STCreturn) && !(p.storageClass & STCreturn)));
+        static uint combine(uint from, uint to) pure nothrow @nogc @safe
+        {
+            return (from << 2) | to;
+        }
```

Comment:
You could move the translation into the combine instead of introducing another enum.
`combine(STCreturn | STCscope, 0)`
The cases are CTFE'ed anyhow.